src/main/java/carpet/settings/ParsedRule.java

```diff
@@ -50,6 +69,11 @@
                 this.validators.add((Validator<T>) callConstructor(Validator._COMMAND_LEVEL_VALIDATOR.class));
             }
         }
+        if (!scarpetApp.isEmpty())
+        {
+            this.hasScarpet = true;
```

Comment:
isn't hasScarpet litttle redundant? should be sufficient to set `this.scarpetApp = null` and use that instead

Reply:
Well, not actually `null`, because it seems that `null` can't be a default for an annotation, but will check for `empty()`.

This was just the way I started because I found `isClient`, but it's true that it's quite redundant, since it's only used in two places.